### PR TITLE
Metric: PieAPP

### LINKED ISSUE #181: Add metric PieAPP
**Is your feature request related to a problem? Please describe.**
Proposed in [PieAPP: Perceptual Image-Error Assessment through Pairwise Preference](https://arxiv.org/pdf/1806.02067.pdf) paper this method showed great performance on [PIPAL dataset](https://arxiv.org/pdf/2007.12142.pdf) which is the largest and most diverse dataset today (as far as I know). 

**Describe the solution you'd like**
Re-implement model using original code, simplify and unify calling interface with DISTS and LPIPS metrics.

**Additional context**
[Code from authors](https://github.com/prashnani/PerceptualImageError)
[Link to model weights](https://web.ece.ucsb.edu/~ekta/projects/PieAPPv0.1/weights/PieAPPv0.1.pth)

### PR DESCRIPTION
**Pull request by zakajd**
Closes #181 

Original code is badly written, so I changed it in a few places. 
Model structure is similar to perceptual losses, but small differences required to overwrite every call, so I decided not to inherit from `ContentLoss`.

## Proposed Changes
- [x] Metric re-implementation with equal performance to original version
- [x] Updated docs
- [x] Add model weights to last release 
- [x] Tests

### CODE REVIEW

**Comment by snk4tr** [comment]
@zakajd thanks a lot for your contribution! I will have a look tomorrow morning.

File: piq/pieapp.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        self.conv4 = nn.Conv2d(features, features * 2, kernel_size=3, padding=1)
+        self.conv5 = nn.Conv2d(features * 2, features * 2, kernel_size=3, padding=1)
+        self.conv6 = nn.Conv2d(features * 2, features * 2, kernel_size=3, padding=1)
+        self.conv7 = nn.Conv2d(features * 2, features * 4, kernel_size=3, padding=1)
+        self.conv8 = nn.Conv2d(features * 4, features * 4, kernel_size=3, padding=1)
+        self.conv9 = nn.Conv2d(features * 4, features * 4, kernel_size=3, padding=1)
+        self.conv10 = nn.Conv2d(features * 4, features * 8, kernel_size=3, padding=1)
+        self.conv11 = nn.Conv2d(features * 8, features * 8, kernel_size=3, padding=1)
+
+        self.fc1_score = nn.Linear(in_features=120832, out_features=512, bias=True)
```
Comment (snk4tr) [inline piq/pieapp.py:43]: I guess this line makes having `features` as a parameter senseless because number of input features is hardcoded. I also think that since you use pretrained model weights you can't really change parameters.

File: piq/pieapp.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        # Term for numerical stability
+        self.EPS = 1e-6
+
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""
+        Forward pass batch of square patches with shape  (N, C, features, features)
+        Returns:
+            
+        """
+        assert x.shape[2] == x.shape[3] == self.features, \
```
Comment (snk4tr) [inline piq/pieapp.py:58]: Since this model is public and can be applied by a user for feature extraction, I also propose to check that the passed tensor meets other requirements (non-negative - if required? 4D etc.)

File: piq/pieapp.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            f"Expected square input with shape {self.features, self.features}"
```
Comment (snk4tr) [inline piq/pieapp.py:59]:
```suggestion
            f"Expected square input with shape {self.features, self.features}, got {x.shape}"
```

File: piq/pieapp.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        self.ref_score_subtract = nn.Linear(in_features=1, out_features=1, bias=True)
+
+        # Term for numerical stability
+        self.EPS = 1e-6
+
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""
+        Forward pass batch of square patches with shape  (N, C, features, features)
+        Returns:
+
```
Comment (snk4tr) [inline piq/pieapp.py:56]: I think this part of the doctoring is missed

File: piq/pieapp.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        self.fc1_weight = nn.Linear(in_features=2048, out_features=512)
+        self.fc2_weight = nn.Linear(in_features=512, out_features=1, bias=True)
+        self.ref_score_subtract = nn.Linear(in_features=1, out_features=1, bias=True)
+
+        # Term for numerical stability
+        self.EPS = 1e-6
+
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""
+        Forward pass batch of square patches with shape  (N, C, features, features)
```
Comment (snk4tr) [inline piq/pieapp.py:54]: Not sure but seems like it should be this way:
```suggestion
        Forward pass a batch of square patches with shape  (N, C, self.features, self.features)
```

File: piq/pieapp.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        x11 = self.flatten(F.relu(self.conv11(self.pool(F.relu(self.conv10(x9))))))
+        # flatten and concatenate
+        features = torch.cat((self.flatten(x3), self.flatten(x5), self.flatten(x7), self.flatten(x9), x11), dim=1)
+        return features, x11
+
+    def compute_difference(self, features_diff, weights_diff):
+        r"""
+        Args:
+            features_diff: Tensor of shape (N * NUM_PATCHES, C_1)
+            features_diff: Tensor of shape (N * NUM_PATCHES, C_2)
```
Comment (snk4tr) [inline piq/pieapp.py:79]: What is `NUM_PATCHES`? I don't see this global variable in the module

File: piq/pieapp.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        self.model.to(device=prediction.device)
+        prediction_features, prediction_weights = self.get_features(prediction)
+        target_features, target_weights = self.get_features(target)
+
+        distances, weights = self.model.compute_difference(
+            target_features - prediction_features,
+            target_weights - prediction_weights
+        )
+
+        # Shape (N, NUM_PATCHES)
```
Comment (snk4tr) [inline piq/pieapp.py:167]: What is `NUM_PATCHES`? I don't see this variable in the module

File: piq/pieapp.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+class PieAPP(_Loss):
+    r"""
+    Implementation of Perceptual Image-Error Assessment through Pairwise Preference.
+    
+    Expects input to be in range [0, 1] with no normalization and RGB channel order.
+    Input images are croped into smaller patches and final score is mean of patch scores.
+    
+
+    Args:
+        reduction: Reduction over samples in batch: "mean"|"sum"|"none".
```
Comment (snk4tr) [inline piq/pieapp.py:101]: If `final score is mean of patch scores`, then I think `reduction` parameters is a bit redundant. We either should avoid explicit description in the doctoring (which would be weird if this is a part of an algorithm for the metric) or remove other redundancy options. What do you think?

File: piq/pieapp.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        
+        Returns:
+            features: List of features extracted from intermediate layers
+        """
+        # Rescale to [0, 255] range on which models was trained
+        x = (x / float(self.data_range)) * 255
+
+        x_patches = crop_patches(x, size=64, stride=self.stride)
+
+        features, weights = self.model(x_patches)
```
Comment (snk4tr) [inline piq/pieapp.py:194]: What do you think about adding this line? May be useful when PieAPP is used as metric.
```suggestion
        with torch.no_grad():
            features, weights = self.model(x_patches)
```

File: piq/functional/base.py [hunk 10]
```diff
@@ -84,3 +84,16 @@ def pow_for_complex(base: torch.Tensor, exp: Union[int, float]) -> torch.Tensor:
     x_real_pow = x_complex_pow_r * torch.cos(x_complex_pow_phi)
     x_imag_pow = x_complex_pow_r * torch.sin(x_complex_pow_phi)
     return torch.stack((x_real_pow, x_imag_pow), dim=-1)
+
+
+def crop_patches(x: torch.Tensor, size=64, stride=32) -> torch.Tensor:
```
Comment (snk4tr) [inline piq/functional/base.py:89]: What happens if size < x[-1] (or x[-2])?  
I think we need a test to cover this case.

**Review by snk4tr — CHANGES_REQUESTED** [review]
According to the codecov, `replace_pooling` functionality is not covered with tests. Could you add a test for the `PieAPPModel` that would verify that layers are replaces and model still works after this change?

Overall, this is a great PR! Only minor changes are required.

File: piq/pieapp.py (hunk 9 above)
Comment (zakajd) [reply piq/pieapp.py:194]: Added `torch.autograd.set_grad_enabled` context manager and additional flag in `init`. 
Some users may want to use PieAPP as a loss, so turning off all gradients isn't good idea.

File: piq/pieapp.py (hunk 8 above)
Comment (zakajd) [reply piq/pieapp.py:101]: Updated docstring description. Final score for _individual_ image  is mean of it's patch scores. So we still may want to reduce results over batch.

File: piq/pieapp.py (hunk 7 above)
Comment (zakajd) [reply piq/pieapp.py:167]: Removed this comment. 
Number of patches depend on input image size, crop size and stride, it's not some fixed value.

File: piq/pieapp.py (hunk 1 above)
Comment (zakajd) [reply piq/pieapp.py:43]: Agree. Removed it from __init__ and hardcoded for now.
As soon as dataset used for PieAPP model training will be released we can reconsider this implementation

File: piq/pieapp.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+        enable_grad: Flag to compute gradients. Usefull when PieAPP used as a loss. Default: False.
+    
+    References:
+        .. [1] Ekta Prashnani, Hong Cai, Yasamin Mostofi, Pradeep Sen
+            (2018). PieAPP: Perceptual Image-Error Assessment through Pairwise Preference
+            https://arxiv.org/abs/1806.02067
+        .. [2] https://github.com/prashnani/PerceptualImageError
+
+    """
+    # TODO(zakajd) 10/12/2020: Load weights to release and change this link
```
Comment (zakajd) [inline piq/pieapp.py:117]: @snk4tr this PR is ready for merge if you don't have additional comments.

I'm currently in the place with very slow internet, so it would be great if you also upload model weights to last release and change the link in code.

File: piq/pieapp.py (hunk 1 above)
Comment (snk4tr) [reply piq/pieapp.py:46]: I also added a TODO comment to not forget

File: piq/pieapp.py (hunk 11 above)
Comment (snk4tr) [reply piq/pieapp.py]: Done

**Review by snk4tr — APPROVED** [review]
Nicely done, @zakajd!
